### DISCUSSION TITLE
Is it possible to download provisioning profiles using Spaceship with AppStoreConnect API?

### DISCUSSION THREAD

**Discussion by sharkusmarkus** [fastlane/fastlane#17907] [Q&A]
Quite some time ago I got some excellent help on how to use Spaceship to download provisioning profiles (https://github.com/fastlane/fastlane/issues/5262#issuecomment-230101215) 

It's been working rather well on our CI (Jenkins). The only issue is that 2FA was enabled on the account that the code uses, and so I've had to use the workaround(s) to save the session cookie, which at some point needs updating, so that's currently a manual action. I do not want to turn of 2FA for the account, and it seems that soon 2FA will be a requirement.

I'm therefore looking at the AppStoreConnect API.

What I'd like to know is wether it is possible to adapt my existing code for Spaceship to use the AppStoreConnect API, and if so, could someone tell me what the new commands are that I should be using?

Here's the code I'm currently using:

```
#!/usr/bin/env ruby
# This script will use Spaceship - part of the Fastlane tools - to login to the developer portal and download the specified provisioning profile
the_user_name = ARGV[0]
the_user_password = ARGV[1]
the_team_id = ARGV[2]
the_profile_name = ARGV[3]
base_profile_path = ARGV[4]


require "spaceship"
begin
    puts ""
    # Login
    puts "   Logging into Developer Portal as: '#{the_user_name}'"
    puts ""
    Spaceship.login(the_user_name, the_user_password)

    # Select team
    puts "   Selecting team: '#{the_team_id}'"
    puts ""
    Spaceship.client.team_id = the_team_id

    # Find the exact profile we're looking for
    puts "   Finding profile named: '#{the_profile_name}'"
    puts ""
    my_profile = Spaceship.provisioning_profile.all.find { |p| p.name == the_profile_name }

    # Download the profile
    the_profile_name = the_profile_name.tr(' ', '_')
    the_profile_path = "#{base_profile_path}#{the_profile_name}.mobileprovision"
    puts "   Downloading profile named named: '#{the_profile_name}' to: '#{the_profile_path}'"
    puts ""
    File.write(the_profile_path, my_profile.download)
end
```

I know I need to amend:
`Spaceship.login(the_user_name, the_user_password)`

 to something like:

```
token = Spaceship::ConnectAPI::Token.create(
		key_id: 'the_private_key_id',
		issuer_id: 'the_issuer_id',
		filepath:  File.absolute_path("#{the_path_to_the_key_id_file}")
	)

Spaceship::ConnectAPI.token = token
```

It's the commands for finding and downloading the profile that I'm not so sure about, so any help in that direction would be greatly appreciated! Thanks.

**Comment by markreynoso** [comment]
Yup, the API is pretty limited, but you can create/download your provisioning profiles. 
Take a look at [Sigh](https://github.com/fastlane/fastlane/blob/master/sigh/lib/sigh/runner.rb). Most of the commands you'll need are in there.

**Reply by sharkusmarkus** [reply]
Thank you. I'll go and see how Sigh does it. 
I did actually delve into how to use the API and did come up with something that echos what I was using Spaceship for, so I do have something in place, but it'd be nice to use fastlane for this too.
